src/algebra/category/Module/adjunctions.lean

```diff
@@ -341,3 +341,70 @@ ext R (α.trans (embedding_lift_iso R F).symm)
 end Free
 
 end category_theory
+
+namespace change_of_rings
+
+universes u₁ u₂
+
+namespace restriction_of_scalars
+
+variables {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S)
+variable (M : Module S)
+
+/--Any `S`-module M is also an `R`-module via a ring homomorphism `f : R ⟶ S` by defining
+`r • m := f r • m` (`module.comp_hom`). This is called restriction of scalars. -/
+def obj' : Module R :=
+{ carrier := M,
+  is_add_comm_group := infer_instance,
+  is_module := module.comp_hom M f }
+
+section
+
+include f
+
+/--The `R`-scalar multiplication on `S`-module M defined by `r • m := f r • m`
+-/
+protected def has_smul : has_smul R M :=
+begin
+  haveI : module R M := (obj' f M).is_module,
+  apply_instance
+end
+
+end
+
+localized "notation r ` r•[` f `] ` m :=
+  @@has_smul.smul (restriction_of_scalars.has_smul f _) r m"
+  in change_of_rings
+
+@[simp] lemma smul_def (r : R) (m : M) :
+  (r r•[f] m) = f r • m := rfl
```

Comment:
I'm not an expert at this stuff, but it looks like the notation works better (in that you don't need the parentheses) if you do it like this:
```suggestion
localized "notation r ` r•[` f `] ` :=
  @@has_smul.smul (restriction_of_scalars.has_smul f _) r"
  in change_of_rings

@[simp] lemma smul_def (r : R) (m : M) : r r•[f] m = f r • m := rfl
```